File: algorithms/0003.longest-substring-without-repeating-characters/solution.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
  curr = []
  longest = 0

  s.each_char do |c|
    if curr.include?(c)
      idx = curr.index(c)
      curr = curr[(idx + 1)..-1]
    end

    curr.push(c)
    longest = [longest, curr.size].max
  end

  longest
end

File: algorithms/0003.longest-substring-without-repeating-characters/solution.go
func lengthOfLongestSubstring(s string) int {
	curr := []rune{}
	longest := 0

	for _, r := range s {
		for i := 0; i < len(curr); i++ {
			if curr[i] == r {
				curr = curr[i+1:]
			}
		}

		curr = append(curr, r)
		if len(curr) > longest {
			longest = len(curr)
		}
	}

	return longest
}
